### PR TITLE
templates: Add more string methods

### PR DESCRIPTION
**Pull request by Dr-Emann**
Add starts_with/ends_with/remove_prefix/remove_suffix/substr methods to string when templating.

**Note: This also adds negative numbers to the template grammar**

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/template.pest [hunk 1]
```diff
@@ -25,7 +25,7 @@ raw_literal = @{ literal_char+ }
 literal = { "\"" ~ (raw_literal | escape)* ~ "\"" }
 
 integer_literal = {
-  ASCII_NONZERO_DIGIT ~ ASCII_DIGIT*
+  "-"? ~ ASCII_NONZERO_DIGIT ~ ASCII_DIGIT*
   | "0"
 }
```
Comment (martinvonz) [inline cli/src/template.pest:30]: I would move this to it's own commit. That commit should also test that the places we currently accept integers fail in reasonable ways if you pass in a negative integer.  For example, before this change, we do this:

```
$ jj log -r @ -T 'change_id.short(-10)'
Error: Failed to parse template:  --> 1:17
  |
1 | change_id.short(-10)
  |                 ^---
  |
  = expected template
```

What's the new behavior? I suspect we'll want to update all the existing users of `expect_integer_expression()` to call some new `expect_positive_integer_expression()` helper.

File: cli/tests/test_templater.rs [hunk 2]
```diff
@@ -356,6 +374,32 @@ fn test_templater_string_method() {
...
+    insta::assert_snapshot!(render(r#""wip: testing".remove_prefix("wip: ")"#), @"testing");
+
+    insta::assert_snapshot!(render(r#""bar@my.example.com".remove_suffix("@other.example.com")"#), @"bar@my.example.com");
+    insta::assert_snapshot!(render(r#""bar@other.example.com".remove_suffix("@other.example.com")"#), @"bar");
+
+    insta::assert_snapshot!(render(r#""foo".substr(0, 0)"#), @"");
+    insta::assert_snapshot!(render(r#""foo".substr(0, 1)"#), @"f");
+    insta::assert_snapshot!(render(r#""foo".substr(0, 99)"#), @"foo");
+    insta::assert_snapshot!(render(r#""abcdef".substr(2, -1)"#), @"cde");
+    insta::assert_snapshot!(render(r#""abcdef".substr(-3, 99)"#), @"def");
```
Comment (martinvonz) [inline cli/tests/test_templater.rs:402]: `"abcdef".substr(4, 2)` and/or `"abcdef".substr(-2, -4)` might also be useful to test (it looks like it's handled correctly in the code, but would be good to exercise that code path)

**Review by martinvonz — COMMENTED** [review]
Thanks! This looks useful

File: cli/src/template.pest [hunk 3]
```diff
@@ -25,7 +25,7 @@ raw_literal = @{ literal_char+ }
 literal = { "\"" ~ (raw_literal | escape)* ~ "\"" }
 
 integer_literal = {
-  ASCII_NONZERO_DIGIT ~ ASCII_DIGIT*
+  "-"? ~ ASCII_NONZERO_DIGIT ~ ASCII_DIGIT*
   | "0"
```
Comment (yuja) [inline cli/src/template.pest:29]: nit: I think `-0` is also valid.

(We might want a negate operator at some point, but I don't think it's needed right now. Parsing as a literal should be good enough.)

File: cli/src/template.pest (hunk 1 above)
Comment (yuja) [reply cli/src/template.pest:30]: It will fall back to default or clamped at 0, I think. We'll need a runtime error to report this kind of errors properly.

Anyway, adding tests will help.

File: cli/src/template_builder.rs [hunk 4]
```diff
@@ -329,6 +329,59 @@ fn build_string_method<'a, L: TemplateLanguage<'a>>(
...
+            let needle_property = expect_plain_text_expression(language, build_ctx, needle_node)?;
+            language.wrap_boolean(TemplateFunction::new(
+                (self_property, needle_property),
+                move |(haystack, needle)| {
+                    if starts_with {
+                        haystack.starts_with(&needle)
+                    } else {
+                        haystack.ends_with(&needle)
+                    }
+                },
```
Comment (yuja) [inline cli/src/template_builder.rs:344]: nit: I slightly prefer separate match arm per function name. It will help if we want to build a HashMap of functions to suggest similar functions on typo.

File: cli/src/template_builder.rs [hunk 5]
```diff
@@ -329,6 +329,59 @@ fn build_string_method<'a, L: TemplateLanguage<'a>>(
...
+        }
+        "substr" => {
+            let [start_idx, end_idx] = template_parser::expect_exact_arguments(function)?;
+            let start_idx_property = expect_integer_expression(language, build_ctx, start_idx)?;
+            let end_idx_property = expect_integer_expression(language, build_ctx, end_idx)?;
+            language.wrap_string(TemplateFunction::new(
+                (self_property, start_idx_property, end_idx_property),
+                |(s, start_idx, end_idx)| {
+                    let to_idx =
+                        |i: i64| -> usize { (if i < 0 { s.len() as i64 + i } else { i }) as usize };
```
Comment (yuja) [inline cli/src/template_builder.rs:371]: better to `.try_into()` and clamp instead of wrapping?

**Comment by Dr-Emann** [comment]
I also had wanted to do a `"replace me".replace("(.*) me", "${1}d")` = `"replaced"`, but I wasn't sure how to handle an invalid regex string without panicking.

File: cli/src/template.pest (hunk 1 above)
Comment (Dr-Emann) [reply cli/src/template.pest]: Added tests, looks like the existing functions that take numerics all did what _I_ expected when passed a negative number.

File: CHANGELOG.md [hunk 6]
```diff
@@ -102,6 +102,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Revsets gained a new function `mine()` that aliases `author(exact:"your_email")`.
 
 * `jj log` timestamp format now accepts `.utc()` to convert a timestamp to UTC.
+ 
+* templates now support additional string methods `.begins_with(x)`, `.ends_with(x)`
```
Comment (yuja) [inline CHANGELOG.md:106]: typo: `begins_with` -> `starts_with`

**Review by yuja — APPROVED** [review]
LGTM, thanks!

**Comment by yuja** [comment]
> I also had wanted to do a `"replace me".replace("(.*) me", "${1}d")` = `"replaced"`, but I wasn't sure how to handle an invalid regex string without panicking.

For literal regex, we can parse it early. We'll probably want to cache the parsed regex even if we had a runtime error support.
https://github.com/martinvonz/jj/blob/0dcd2fa26519ce5905dd736be64954ca09987db0/cli/src/template_builder.rs#L437-L446

FWIW, `.replace()` can also support `<kind>:<pattern>` syntax like revset, but I don't know what `exact:` (or `substring:`) should mean in this context.

**Review by martinvonz — APPROVED** [review]
Thanks for adding those tests. I'm glad everything already was able to handle negative numbers. I wouldn't have been surprised if something had panicked.
